File: 25-State/src/go/gumball_machine.go
package main

import (
	"fmt"
)

// GumballMachine is the Context — it owns the states, delegates actions to the
// current one, and holds the shared gumball count.
type GumballMachine struct {
	noQuarter  State
	hasQuarter State
	sold       State
	soldOut    State

	current State
	count   int
}

func NewGumballMachine(count int) *GumballMachine {
	m := &GumballMachine{count: count}
	m.noQuarter = &NoQuarterState{machine: m}
	m.hasQuarter = &HasQuarterState{machine: m}
	m.sold = &SoldState{machine: m}
	m.soldOut = &SoldOutState{machine: m}
	if count > 0 {
		m.current = m.noQuarter
	} else {
		m.current = m.soldOut
	}
	return m
}

func (m *GumballMachine) InsertQuarter() { m.current.InsertQuarter() }
func (m *GumballMachine) EjectQuarter()  { m.current.EjectQuarter() }
func (m *GumballMachine) TurnCrank() {
	m.current.TurnCrank()
	m.current.Dispense()
}

func (m *GumballMachine) setState(state State) { m.current = state }

func (m *GumballMachine) releaseBall() {
	if m.count > 0 {
		fmt.Println("  A gumball rolls out...")
		m.count--
	}
}

func (m *GumballMachine) Status() string {
	return fmt.Sprintf("[state=%s, gumballs=%d]", m.current.Name(), m.count)
}

File: 25-State/src/rust/gumball_machine.rs
use crate::state::State;
use crate::no_quarter_state::NoQuarterState;
use crate::sold_out_state::SoldOutState;

/// Context — owns the current state (as a boxed trait object) and the shared
/// count. Each action takes the state out, runs it, and stores the returned
/// next state.
pub struct GumballMachine {
    state: Option<Box<dyn State>>,
    count: u32,
}

impl GumballMachine {
    pub fn new(count: u32) -> Self {
        let state: Box<dyn State> = if count > 0 {
            Box::new(NoQuarterState)
        } else {
            Box::new(SoldOutState)
        };
        GumballMachine { state: Some(state), count }
    }

    pub fn insert_quarter(&mut self) {
        let state = self.state.take().unwrap();
        self.state = Some(state.insert_quarter(self));
    }

    pub fn eject_quarter(&mut self) {
        let state = self.state.take().unwrap();
        self.state = Some(state.eject_quarter(self));
    }

    pub fn turn_crank(&mut self) {
        let state = self.state.take().unwrap();
        let state = state.turn_crank(self);
        self.state = Some(state.dispense(self));
    }

    pub fn count(&self) -> u32 {
        self.count
    }

    pub fn release_ball(&mut self) {
        if self.count > 0 {
            println!("  A gumball rolls out...");
            self.count -= 1;
        }
    }

    pub fn status(&self) -> String {
        format!("[state={}, gumballs={}]", self.state.as_ref().unwrap().name(), self.count)
    }
}
